text/0000-ubi-based-stacks.md

```diff
@@ -0,0 +1,181 @@
+JVM/JDK, one would be installed via 'yum' or 'dnf'. 
+
+Until recently, CNCF Buildpacks were unable to effect changes to the Builder or Run image at a system level (eg, you
+could not make modifications that required root from a buildpack, such as installing packages via yum). Recent 
+updates to the Buildpacks project have added 'extensions', these can be considered as special variants of buildpacks 
+that are able to modify the Builder image (and substitute the Run image), via generated Dockerfiles created during 
+the execution of the Extensions. (In the future, the goal is to also allow modification of the Run image via the 
+same mechanism).
+
+Via extensions, it becomes possible to have a Builder image that can install dependencies such as Java, Maven,
+NodeJS, etc via yum. These extensions are very similar to buildpacks, participating in the 'detect' phase, and 
+being allowed to issue 'provides' to the buildplan (but critically not 'requires'). Then extensions that are
+selected to become part of a build run a 'generate' task (rather than a buildpacks 'build' task), that may output 
+Dockerfiles to modify the Builder image, and/or substitute the Run image. Just like buildpacks, the detect & 
+generate invocations get access to the source project to determine what actions they wish to take. For more 
+information on extensions see.. https://github.com/buildpacks/spec/blob/main/image_extension.md
+
+### Initial offered image
+
+The intent is provide an initial builder image that caters to Java & NodeJS runtimes (without Java Native). 
+Some early prototyping has shown this to be possible using the existing Java & NodeJS builder ecosystems 
+present within Paketo today, with Extensions installing the main runtime dependencies. 
+
+In the case of Java, much work has already been done by Paketo to allow for different buildpacks to supply the 
+JDK/JVM, making it relatively simple to extend this to allow an Extension to perform the same role. For NodeJS, there
+were some small changes required, that have already been merged. 
+
+The initial image will be labelled with a Stack ID of `io.buildpacks.stacks.ubi8` Note that much as Ubuntu
+moves from release to release, so does UBI, and currently we will target UBI8, but want to ensure the naming allows
+for the future addition for UBI9 and so on.
+
+There will be no "intermediate" image as with Bionic, as the base builder image is intended to be a published
+UBI8 image.
+
+The initial image shall be considered as the 'base' variant, for the purposes of differentiation when/if 
+subsequent images are published. 
+
+### Image Naming and Tagging
+
+The stack will name and tag the release images with the following pattern:
+
+```
+paketobuildpacks/{phase}-ubi8-{variant}:{version}
+```
+
+For example we could see the following images for UBI8 stacks:
+
+* `paketobuildpacks/build-ubi8-base:latest`
+* `paketobuildpacks/run-java-ubi8-base:1.2.3`
+* `paketobuildpacks/run-nodejs-ubi8-base:1.2.3`
```

Comment:
I think I'm misunderstanding - what are the runtime-specific run images needed for, if the extensions will be providing the runtime installations?

Reply:
Because at least in this incarnation, the extensions are only modifying the builder image, not the run image. At present upstream only offers support for builder image modification and run image switching. This is enough to create a viable solution, but does have the downside of requiring pre-canned runtime images. 

Once upstream offers run image modification, then we can offer a solution based around that, but run image modification comes with the cost that it will disable rebasing for any applications that are built using it. 

Run image switching doesn't come with that drawback, so in some ways provides the end user with a better experience at the cost of requiring a greater number of run images. 

The base for each runtime run image already exists as a  published UBI runtime image, except each is lacking the required env vars and labels for it to qualify as a buildpack run image. We are looking at if we can get those added to the upstream images, which would remove the need for Paketo to publish them, but I didn't want to rely on that happening for this RFC, so included the info on Paketo publishing the run images.

Reply:
@barDweller thank you for the added explanation, that cleared up my confusion. 👍